Source/DataModels/Batch/GeometryBatch.h

```diff
@@ -0,0 +1,176 @@
+#include <vector>
+#include <unordered_map>
+
+#include "GL/glew.h"
+
+#define DOUBLE_BUFFERS 2
+
+class ComponentMeshRenderer;
+class ResourceMesh;
+class ResourceMaterial;
+class GameObject;
+class Program;
+
+class GeometryBatch
+{
+public:
+	GeometryBatch(int flag);
+	~GeometryBatch();
+
+	void CreateVAO();
+	void ClearBuffer();
+	bool CleanUp();
+
+	void UpdateBatchComponents();
+
+	void AddComponentMeshRenderer(ComponentMeshRenderer* newComponent);
+
+	void DeleteComponent(ComponentMeshRenderer* componentToDelete);
+	void DeleteMaterial(const ComponentMeshRenderer* componentToDelete);
+	std::vector< ComponentMeshRenderer*> ChangeBatch(const ComponentMeshRenderer* componentToDelete);
+
+	void BindBatch(bool selected);
+	void FillMaterial();
+	void ReserveModelSpace();
+
+	bool IsEmpty() const;
+	bool IsDirty() const;
+
+	void SetFillMaterials(const bool fillMaterials);
+	void SortByDistanceFarToClose();
+	void SortByDistanceCloseToFar();
+	void SetDirty(const bool dirty);
+
+	const int GetFlags() const;
+
+private:
+	struct Command
+	{
+		GLuint  count;			// Number of indices in the mesh
+		GLuint  instanceCount;	// Number of instances to render
+		GLuint  firstIndex;		// Index offset in the EBO
+		GLuint  baseVertex;		// Vertex offset in the VBO
+		GLuint  baseInstance;	// Instance Index
+	};
+
+	struct MaterialMetallic {
+		float4 diffuse_color = float4::zero; //0 //16
+		int has_diffuse_map = 0;			 //20 //4
+		int has_normal_map = 0;				 //24 //4
+		int has_metallic_map = 0;			 //36 //4
+		float smoothness = 0;				 //28 //4
+		float metalness = 0;				 //32 //4
+		float normal_strength = 0;			 //16 //4
+		uint64_t diffuse_map;				 //40 //8
+		uint64_t normal_map;				 //48 //8
+		uint64_t metallic_map;				 //56 //8 -->64
+	};
+
+	struct MaterialSpecular {
+		float4 diffuse_color = float4::zero;  //0  //16
+		float3 specular_color = float3::zero; //16 //16       
+		int has_diffuse_map = 0;              //32 //4
+		int has_normal_map = 0;               //36 //4
+		int has_specular_map = 0;             //40 //4
+		float smoothness = 0;                 //44 //4
+		float normal_strength = 0;            //48 //4
+		uint64_t diffuse_map;				  //48 //8
+		uint64_t normal_map;				  //56 //8
+		uint64_t specular_map;				  //64 //8    
+		float2 padding = float2::zero;		  //72 //8 --> 80
+	};
+
+	struct ResourceInfo
+	{
+		ResourceMesh* resourceMesh;
+		unsigned int vertexOffset;
+		unsigned int indexOffset;
+	};
+
+private:
+	void FillBuffers();
+	void FillEBO();
+
+	void CreateInstanceResourceMesh(ResourceMesh* mesh);
+	int CreateInstanceResourceMaterial(const std::shared_ptr<ResourceMaterial> material);
+
+	ResourceInfo* FindResourceInfo(const ResourceMesh* mesh);
+
+	void LockBuffer();
+	void WaitBuffer();
+
+	std::vector<ComponentMeshRenderer*> componentsInBatch;
+	std::vector<std::shared_ptr<ResourceMaterial>> resourcesMaterial;
+	std::unordered_map<const ComponentMeshRenderer*, int>  objectIndexes;
+	std::vector<ResourceInfo*> resourcesInfo;
```

Comment:
Any reason this is a vector of pointers instead of just a vector of references? Could be wrong, but I think we don't need to use pointers for this.
```suggestion
	std::vector<ResourceInfo> resourcesInfo;
```
If we do need to use pointers, any reason for not making them unique?

Reply:
It was done with a pointer in case it was needed to be re-assigned during the development or the future, while if it were with a reference it could only be assigned at initialization. 
For the unique, the same as before, we prefer using raw pointers and they are being deleted and the vectors cleared so i don't see the need for using unique.